src/bucket/BucketManagerImpl.cpp

```diff
@@ -310,6 +365,41 @@ BucketManagerImpl::getBucketByHash(uint256 const& hash)
     return std::shared_ptr<Bucket>();
 }
 
+std::shared_future<std::shared_ptr<Bucket>>
+BucketManagerImpl::getMergeFuture(MergeKey const& key)
+{
+    std::lock_guard<std::recursive_mutex> lock(mBucketMutex);
+    auto i = mLiveFutures.find(key);
+    if (i == mLiveFutures.end())
+    {
+        // If there's no live future, we might be able to _make_
+        // one for a retained bucket, if we still know its inputs.
+        auto j = mSharedBucketMergesByMergeKey.find(key);
+        if (j != mSharedBucketMergesByMergeKey.end())
+        {
+            auto bucket = getBucketByHash(j->second);
+            if (bucket)
+            {
+                std::promise<std::shared_ptr<Bucket>> promise;
+                auto future = promise.get_future().share();
+                promise.set_value(bucket);
+                mLiveFutures.insert(std::make_pair(key, future));
+                return future;
+            }
+        }
+        return std::shared_future<std::shared_ptr<Bucket>>();
+    }
+    return i->second;
+}
+
+void
+BucketManagerImpl::putMergeFuture(
+    MergeKey const& key, std::shared_future<std::shared_ptr<Bucket>> wp)
+{
+    std::lock_guard<std::recursive_mutex> lock(mBucketMutex);
+    mLiveFutures.emplace(std::make_pair(key, wp));
```

Comment:
emplace doesn't need the extra `std::make_pair`

Reply:
fixed